### PR TITLE
raft: separate apply from ready handling

### PR DESCRIPTION
**Pull request by BusyJay**
When implementing async apply, we have to update apply index after async apply is actually finished, otherwise a follower with wrong quorum may become leader in some rare cases. This pr tries to separate Ready into appending and applying, so we can update apply index only when it has to.

@siddontang @hhkbp2 PTAL

### CODE REVIEW

**Comment by siddontang** [comment]
@BusyJay 

Can you describe more clearly and give an example for the `a follower with wrong quorum may become leader in some rare cases`. Maybe https://github.com/pingcap/tikv/pull/1488#discussion_r96134636? 

Seem that etcd may also have this problem /cc @xiang90

**Comment by siddontang** [comment]
Any test to cover the applied_index is not none?

File: src/raft/raw_node.rs [hunk 1]
```diff
@@ -291,14 +291,19 @@ impl<T: Storage> RawNode<T> {
...
+                   Some(applied_idx))
+    }
+
     // Ready returns the current point-in-time state of this RawNode.
     pub fn ready(&mut self) -> Ready {
-        Ready::new(&mut self.raft, &self.prev_ss, &self.prev_hs)
+        Ready::new(&mut self.raft, &self.prev_ss, &self.prev_hs, None)
     }
 
-    // HasReady called when RawNode user need to check if any Ready pending.
```
Comment (hhkbp2) [inline src/raft/raw_node.rs]: Comments should be kept.

**Comment by BusyJay** [comment]
Considering a cluster with 3 nodes, A, B, C. A is leader. And A add two nodes D, E to the cluster. C has committed these two entries but not applied yet. So if for some reason C is going to start an election, it will calculate the quorum as `3 / 2 + 1 = 2`. But the actual quorum should be `5 / 2 + 1 = 3`.

Please note that Line https://github.com/pingcap/tikv/blob/master/src/raft/raft.rs#L870 is used to prevent a peer from starting a new election in the similar case by checking if there is any confchange in the pending apply entries. But if we apply the log asynchronously and updating the applied index without waiting the apply actually finished, the check will be bypassed.

File: src/raft/raw_node.rs (hunk 1 above)
Comment (BusyJay) [inline src/raft/raw_node.rs]: The comment is kept as L334.

**Comment by siddontang** [comment]
Seem that etcd uses `pendingConf` to avoid this. The `pendingConf` can be reset after the ConfChange applied. 

/cc @xiang90

**Comment by BusyJay** [comment]
No, `pendingConf` is not updated when the node is follower.

**Comment by xiang90** [comment]
Once async apply is enabled, the candidate must apply all pending committed entries before it can send out the vote requests at application layer.

**Comment by siddontang** [comment]
Hi @xiang90 

> the candidate must apply all pending committed entries before it can send out the vote requests at application layer

Could you show me the code reference?

**Comment by siddontang** [comment]
PTAL @hhkbp2

**Comment by hhkbp2** [comment]
LGTM
Do we need to update these code etcd/raft?

**Comment by siddontang** [comment]
No need now. I think etcd will fix it in another way.

**Pull request merged**